src/plugins/intel_gpu/src/graph/primitive_inst.cpp

```diff
@@ -638,76 +655,95 @@ event::ptr primitive_inst::realloc_if_needed() {
-                               << "Reset memory" << std::endl;
-        _max_output_layout_count = 0;
+    for (size_t i = 0; i < updated_layouts.size(); ++i) {
+        if (updated_layouts[i].get_buffer_size().count() * 10 < _max_output_layout_count[i]) {
+            GPU_DEBUG_TRACE_DETAIL << id() << ": Updated output[" << i << "] size " << updated_layouts[i].get_buffer_size().count()
+                                   << " is much smaller than current memory size! " << _max_output_layout_count[i]
+                                   << "Reset memory of output " << i << std::endl;
+            _max_output_layout_count[i] = 0;
+        }
     }
 
-    bool can_reuse_buffer = _outputs[0] && updated_layout.get_buffer_size().count() <= _max_output_layout_count;
     // Handle runtime dynamic concat optimization
     if (_node->is_type<concatenation>() && can_be_optimized() && allocation_done_by_other) {
         allocation_done_by_other = false;
         GPU_DEBUG_PROFILED_STAGE_MEMALLOC_INFO("concat_alloc_by_other");
         return ev;
     }
 
-    std::pair<bool, ov::Shape> prealloc_info;
     int32_t tmp_prealloc_count = get_prealloc_iter_num();
     GPU_DEBUG_IF(debug_config->mem_preallocation_params.is_initialized) {
         // If debug config is set, repsect the config most
         tmp_prealloc_count = -1;
     }
-    prealloc_info = sp.predict_preallocation_shape(id(), updated_layout, can_reuse_buffer, tmp_prealloc_count);
 
-    if (prealloc_info.first && sp.can_preallocate(ov::shape_size(prealloc_info.second) * (dt_size / 8))) {
-        auto new_layout = updated_layout;
-        new_layout.set_partial_shape(prealloc_info.second);
-        updated_params.output_layouts[0] = new_layout;
-    }
-
-    if (updated_params.output_layouts[0].get_buffer_size().count() < updated_layout.get_buffer_size().count())
-        updated_params.output_layouts[0] = updated_layout;
-    if (can_reuse_buffer) {
-        GPU_DEBUG_TRACE_DETAIL << id() << ": reuse previously allocated output buffer - "
-                               << actual_layout.get_buffer_size().count() << "/" << _max_output_layout_count
-                               << std::endl;
-        if (_outputs[0]->get_layout() != actual_layout) {
-            _outputs[0] = _network.get_engine().reinterpret_buffer(*_outputs[0], actual_layout);
+    for (size_t i = 0; i < actual_layouts.size(); ++i) {
+        bool can_reuse_buffer = (_outputs[i] && updated_layouts[i].get_buffer_size().count() <= _max_output_layout_count[i]);
+        std::pair<bool, ov::Shape> prealloc_info =
+            sp.predict_preallocation_shape(id(), updated_layouts[i], can_reuse_buffer, i, tmp_prealloc_count);
+        if (prealloc_info.first && sp.can_preallocate(ov::shape_size(prealloc_info.second) * (dt_sizes[i] / 8))) {
+            auto new_layout = updated_layouts[i];
+            new_layout.set_partial_shape(prealloc_info.second);
+            updated_params.output_layouts[i] = new_layout;
```

Comment:
I think we should simplify that part too (maybe in other PR).
Maybe something like this:
```cpp
updated_params.output_layouts[i] = updated_layouts[i].clone_with_new_shape(prealloc_info.second);
```